In [3]:
import pandas as pd

In [10]:
df = pd.read_csv('https://studio.ironhack.school/asset-v1:IRONHACK+DAFT+202006_PAR+type@asset+block@datasets_674388_1186156_tv_shows.csv')

In [11]:
df

Unnamed: 0                            Title  Year  Age  IMDb  \
0              0                     Breaking Bad  2008  18+   9.5   
1              1                  Stranger Things  2016  16+   8.8   
2              2                      Money Heist  2017  18+   8.4   
3              3                         Sherlock  2010  16+   9.1   
4              4                 Better Call Saul  2015  18+   8.7   
...          ...                              ...   ...  ...   ...   
5606        5606  Tut's Treasures: Hidden Secrets  2018  NaN   NaN   
5607        5607                 Paradise Islands  2017  NaN   NaN   
5608        5608                      Wild Russia  2018  NaN   NaN   
5609        5609                      Love & Vets  2017  NaN   NaN   
5610        5610         United States of Animals  2016  NaN   NaN   

     Rotten Tomatoes  Netflix  Hulu  Prime Video  Disney+  type  
0                96%        1     0            0        0     1  
1                93%        1     0            0        0     1  
2                91%        1     0            0        0     1  
3                78%        1     0            0        0     1  
4                97%        1     0            0        0     1  
...              ...      ...   ...          ...      ...   ...  
5606             NaN        0     0            0        1     1  
5607             NaN        0     0            0        1     1  
5608             NaN        0     0            0        1     1  
5609             NaN        0     0            0        1     1  
5610             NaN        0     0            0        1     1  

[5611 rows x 11 columns]

In [14]:
df = pd.read_csv('https://studio.ironhack.school/asset-v1:IRONHACK+DAFT+202006_PAR+type@asset+block@datasets_674388_1186156_tv_shows.csv', index_col="Unnamed: 0")

In [16]:
df

Title  Year  Age  IMDb Rotten Tomatoes  \
0                        Breaking Bad  2008  18+   9.5             96%   
1                     Stranger Things  2016  16+   8.8             93%   
2                         Money Heist  2017  18+   8.4             91%   
3                            Sherlock  2010  16+   9.1             78%   
4                    Better Call Saul  2015  18+   8.7             97%   
...                               ...   ...  ...   ...             ...   
5606  Tut's Treasures: Hidden Secrets  2018  NaN   NaN             NaN   
5607                 Paradise Islands  2017  NaN   NaN             NaN   
5608                      Wild Russia  2018  NaN   NaN             NaN   
5609                      Love & Vets  2017  NaN   NaN             NaN   
5610         United States of Animals  2016  NaN   NaN             NaN   

      Netflix  Hulu  Prime Video  Disney+  type  
0           1     0            0        0     1  
1           1     0            0        0     1  
2           1     0            0        0     1  
3           1     0            0        0     1  
4           1     0            0        0     1  
...       ...   ...          ...      ...   ...  
5606        0     0            0        1     1  
5607        0     0            0        1     1  
5608        0     0            0        1     1  
5609        0     0            0        1     1  
5610        0     0            0        1     1  

[5611 rows x 10 columns]

In [19]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5611 entries, 0 to 5610
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Title            5611 non-null   object 
 1   Year             5611 non-null   int64  
 2   Age              3165 non-null   object 
 3   IMDb             4450 non-null   float64
 4   Rotten Tomatoes  1011 non-null   object 
 5   Netflix          5611 non-null   int64  
 6   Hulu             5611 non-null   int64  
 7   Prime Video      5611 non-null   int64  
 8   Disney+          5611 non-null   int64  
 9   type             5611 non-null   int64  
dtypes: float64(1), int64(6), object(3)
memory usage: 482.2+ KB


In [21]:
df.columns

Index(['Title', 'Year', 'Age', 'IMDb', 'Rotten Tomatoes', 'Netflix', 'Hulu',
       'Prime Video', 'Disney+', 'type'],
      dtype='object')

In [27]:
df.columns=['Title', 'Year', 'Age', 'IMDb', 'Age_restr',
            'Netflix', 'Hulu', 'Prime Video', 'Disney+', 'type']

In [28]:
df.head()

Title  Year  Age  IMDb Age_restr  Netflix  Hulu  Prime Video  \
0      Breaking Bad  2008  18+   9.5       96%        1     0            0   
1   Stranger Things  2016  16+   8.8       93%        1     0            0   
2       Money Heist  2017  18+   8.4       91%        1     0            0   
3          Sherlock  2010  16+   9.1       78%        1     0            0   
4  Better Call Saul  2015  18+   8.7       97%        1     0            0   

   Disney+  type  
0        0     1  
1        0     1  
2        0     1  
3        0     1  
4        0     1

In [29]:
df.rename(columns={'Age_restr':'Age_restriction'})

Title  Year  Age  IMDb Age_restriction  \
0                        Breaking Bad  2008  18+   9.5             96%   
1                     Stranger Things  2016  16+   8.8             93%   
2                         Money Heist  2017  18+   8.4             91%   
3                            Sherlock  2010  16+   9.1             78%   
4                    Better Call Saul  2015  18+   8.7             97%   
...                               ...   ...  ...   ...             ...   
5606  Tut's Treasures: Hidden Secrets  2018  NaN   NaN             NaN   
5607                 Paradise Islands  2017  NaN   NaN             NaN   
5608                      Wild Russia  2018  NaN   NaN             NaN   
5609                      Love & Vets  2017  NaN   NaN             NaN   
5610         United States of Animals  2016  NaN   NaN             NaN   

      Netflix  Hulu  Prime Video  Disney+  type  
0           1     0            0        0     1  
1           1     0            0        0     1  
2           1     0            0        0     1  
3           1     0            0        0     1  
4           1     0            0        0     1  
...       ...   ...          ...      ...   ...  
5606        0     0            0        1     1  
5607        0     0            0        1     1  
5608        0     0            0        1     1  
5609        0     0            0        1     1  
5610        0     0            0        1     1  

[5611 rows x 10 columns]

In [31]:
df.head()

Title  Year  Age  IMDb Age_restr  Netflix  Hulu  Prime Video  \
0      Breaking Bad  2008  18+   9.5       96%        1     0            0   
1   Stranger Things  2016  16+   8.8       93%        1     0            0   
2       Money Heist  2017  18+   8.4       91%        1     0            0   
3          Sherlock  2010  16+   9.1       78%        1     0            0   
4  Better Call Saul  2015  18+   8.7       97%        1     0            0   

   Disney+  type  
0        0     1  
1        0     1  
2        0     1  
3        0     1  
4        0     1

In [33]:
#1st option:
##df=df.rename(columns={'Age_restr':'Age_restriction'})

#2nd option:
df.rename(columns={'Age_restr':'Age_restriction'}, inplace=True)

In [35]:
df.head()

Title  Year  Age  IMDb Age_restriction  Netflix  Hulu  \
0      Breaking Bad  2008  18+   9.5             96%        1     0   
1   Stranger Things  2016  16+   8.8             93%        1     0   
2       Money Heist  2017  18+   8.4             91%        1     0   
3          Sherlock  2010  16+   9.1             78%        1     0   
4  Better Call Saul  2015  18+   8.7             97%        1     0   

   Prime Video  Disney+  type  
0            0        0     1  
1            0        0     1  
2            0        0     1  
3            0        0     1  
4            0        0     1

## Add a column

In [37]:
df['new_column']=7
df.head()

Title  Year  Age  IMDb Age_restriction  Netflix  Hulu  \
0      Breaking Bad  2008  18+   9.5             96%        1     0   
1   Stranger Things  2016  16+   8.8             93%        1     0   
2       Money Heist  2017  18+   8.4             91%        1     0   
3          Sherlock  2010  16+   9.1             78%        1     0   
4  Better Call Saul  2015  18+   8.7             97%        1     0   

   Prime Video  Disney+  type  new_column  
0            0        0     1           7  
1            0        0     1           7  
2            0        0     1           7  
3            0        0     1           7  
4            0        0     1           7

In [38]:
df["duplicated_column"]=df.Title
df.head()

Title  Year  Age  IMDb Age_restriction  Netflix  Hulu  \
0      Breaking Bad  2008  18+   9.5             96%        1     0   
1   Stranger Things  2016  16+   8.8             93%        1     0   
2       Money Heist  2017  18+   8.4             91%        1     0   
3          Sherlock  2010  16+   9.1             78%        1     0   
4  Better Call Saul  2015  18+   8.7             97%        1     0   

   Prime Video  Disney+  type  new_column duplicated_column  
0            0        0     1           7      Breaking Bad  
1            0        0     1           7   Stranger Things  
2            0        0     1           7       Money Heist  
3            0        0     1           7          Sherlock  
4            0        0     1           7  Better Call Saul

Series with IMDb below 9.2 are too bad for me to watch

In [42]:
# detect titles with IMDb>9.2
#df['IMDb']>9.2

# select the titles with IMDb
#df.loc[df["IMDb"]>9.2]

#assign the 'to watch' value
df.loc[df["IMDb"]>9.2, "todolist"]="to watch"

In [44]:
df.head()

Title  Year  Age  IMDb Age_restriction  Netflix  Hulu  \
0      Breaking Bad  2008  18+   9.5             96%        1     0   
1   Stranger Things  2016  16+   8.8             93%        1     0   
2       Money Heist  2017  18+   8.4             91%        1     0   
3          Sherlock  2010  16+   9.1             78%        1     0   
4  Better Call Saul  2015  18+   8.7             97%        1     0   

   Prime Video  Disney+  type  new_column duplicated_column  todolist  
0            0        0     1           7      Breaking Bad  to watch  
1            0        0     1           7   Stranger Things       NaN  
2            0        0     1           7       Money Heist       NaN  
3            0        0     1           7          Sherlock       NaN  
4            0        0     1           7  Better Call Saul       NaN

In [45]:
df.loc[~(df["IMDb"]>9.2), "todolist"]="skip"
df.head()

Title  Year  Age  IMDb Age_restriction  Netflix  Hulu  \
0      Breaking Bad  2008  18+   9.5             96%        1     0   
1   Stranger Things  2016  16+   8.8             93%        1     0   
2       Money Heist  2017  18+   8.4             91%        1     0   
3          Sherlock  2010  16+   9.1             78%        1     0   
4  Better Call Saul  2015  18+   8.7             97%        1     0   

   Prime Video  Disney+  type  new_column duplicated_column  todolist  
0            0        0     1           7      Breaking Bad  to watch  
1            0        0     1           7   Stranger Things      skip  
2            0        0     1           7       Money Heist      skip  
3            0        0     1           7          Sherlock      skip  
4            0        0     1           7  Better Call Saul      skip

## Unique values

In [47]:
df.nunique()

Title                5564
Year                   81
Age                     5
IMDb                   78
Age_restriction        88
Netflix                 2
Hulu                    2
Prime Video             2
Disney+                 2
type                    1
new_column              1
duplicated_column    5564
todolist                2
dtype: int64

In [48]:
df.Hulu.nunique()

2

In [50]:
len(set(df["Age_restriction"]))

89

In [51]:
df[['Hulu','Disney+']].nunique()

Hulu       2
Disney+    2
dtype: int64

In [52]:
df["Age_restriction"].unique()

array(['96%', '93%', '91%', '78%', '97%', '81%', '83%', '92%', '100%',
       '94%', '89%', '88%', '72%', '67%', '85%', '87%', '90%', '75%',
       '70%', '86%', '71%', '82%', '64%', '95%', '77%', '63%', '99%',
       '50%', '80%', '84%', nan, '98%', '68%', '62%', '40%', '39%', '57%',
       '69%', '54%', '44%', '66%', '65%', '27%', '45%', '37%', '74%',
       '51%', '61%', '73%', '33%', '38%', '11%', '60%', '30%', '28%',
       '55%', '24%', '36%', '32%', '47%', '58%', '35%', '56%', '22%',
       '79%', '42%', '6%', '41%', '19%', '8%', '43%', '17%', '31%', '20%',
       '34%', '76%', '53%', '59%', '52%', '9%', '48%', '23%', '25%',
       '21%', '18%', '46%', '26%', '14%', '15%'], dtype=object)

## Frequency of values

In [53]:
l=[1,2,3,3,2,2,3,23,2,31,23,12,31,23,12,3,12,3,123,12,3,12,3,12,3,2,3,123,12,3,12,3,12,312]
import operator
dict(sorted({i:l.count(i) for i in l}.items(), key=operator.itemgetter(1), reverse=True))

{3: 11, 12: 9, 2: 5, 23: 3, 31: 2, 123: 2, 1: 1, 312: 1}

In [55]:
pd.Series(l).value_counts()

3      11
12      9
2       5
23      3
31      2
123     2
1       1
312     1
dtype: int64

In [58]:
df.todolist.value_counts()

skip        5601
to watch      10
Name: todolist, dtype: int64

# Descriptive statistics and calculations

- Mean
- Median
- Mode
- Border values (minimum and maximum)
- Percentiles
- Standard Deviation (variance)
- Support(number of rows)

In [59]:
df.mean()

Year           2011.021030
IMDb              7.113258
Netflix           0.344145
Hulu              0.312600
Prime Video       0.382107
Disney+           0.032080
type              1.000000
new_column        7.000000
dtype: float64

In [60]:
df.mean(axis=1)

0       253.312500
1       254.225000
2       254.300000
3       253.512500
4       254.087500
           ...    
5606    289.571429
5607    289.428571
5608    289.571429
5609    289.428571
5610    289.285714
Length: 5611, dtype: float64

In [62]:
df.median()

Year           2015.0
IMDb              7.3
Netflix           0.0
Hulu              0.0
Prime Video       0.0
Disney+           0.0
type              1.0
new_column        7.0
dtype: float64

In [64]:
df['Age_restriction'].mode()

0    100%
dtype: object

In [66]:
df['Hulu'].mode()

0    0
dtype: int64

In [67]:
df.Title.mode()

0       Kingdom
1    Undercover
dtype: object

In [68]:
df.Title.value_counts()

Kingdom                             3
Undercover                          3
Rome                                2
Elizabeth I                         2
Fruits Basket                       2
                                   ..
Dirty Dancing                       1
Conviction                          1
Tayo the Little Bus                 1
Spartan: Ultimate Team Challenge    1
Utawarerumono: The False Faces      1
Name: Title, Length: 5564, dtype: int64

In [69]:
df.mode()

Title    Year  Age  IMDb Age_restriction  Netflix  Hulu  Prime Video  \
0     Kingdom  2017.0  16+   7.4            100%      0.0   0.0          0.0   
1  Undercover     NaN  NaN   NaN             NaN      NaN   NaN          NaN   

   Disney+  type  new_column duplicated_column todolist  
0      0.0   1.0         7.0           Kingdom     skip  
1      NaN   NaN         NaN        Undercover      NaN

In [71]:
df=df.rename(columns={'Age_restriction':'Rotten Tomatoes'})
df.head()

Title  Year  Age  IMDb Rotten Tomatoes  Netflix  Hulu  \
0      Breaking Bad  2008  18+   9.5             96%        1     0   
1   Stranger Things  2016  16+   8.8             93%        1     0   
2       Money Heist  2017  18+   8.4             91%        1     0   
3          Sherlock  2010  16+   9.1             78%        1     0   
4  Better Call Saul  2015  18+   8.7             97%        1     0   

   Prime Video  Disney+  type  new_column duplicated_column  todolist  
0            0        0     1           7      Breaking Bad  to watch  
1            0        0     1           7   Stranger Things      skip  
2            0        0     1           7       Money Heist      skip  
3            0        0     1           7          Sherlock      skip  
4            0        0     1           7  Better Call Saul      skip

In [72]:
df.mode()

Title    Year  Age  IMDb Rotten Tomatoes  Netflix  Hulu  Prime Video  \
0     Kingdom  2017.0  16+   7.4            100%      0.0   0.0          0.0   
1  Undercover     NaN  NaN   NaN             NaN      NaN   NaN          NaN   

   Disney+  type  new_column duplicated_column todolist  
0      0.0   1.0         7.0           Kingdom     skip  
1      NaN   NaN         NaN        Undercover      NaN

In [73]:
df.min()

Title                #MeToo, Now What?
Year                              1901
IMDb                               1.0
Netflix                              0
Hulu                                 0
Prime Video                          0
Disney+                              0
type                                 1
new_column                           7
duplicated_column    #MeToo, Now What?
todolist                          skip
dtype: object

In [74]:
df.max()

Title                頭文字D First Stage
Year                             2020
IMDb                              9.6
Netflix                             1
Hulu                                1
Prime Video                         1
Disney+                             1
type                                1
new_column                          7
duplicated_column    頭文字D First Stage
todolist                     to watch
dtype: object

In [76]:
df.quantile(q=0.75)

Year           2017.0
IMDb              7.9
Netflix           1.0
Hulu              1.0
Prime Video       1.0
Disney+           0.0
type              1.0
new_column        7.0
Name: 0.75, dtype: float64

In [77]:
# Median
df.quantile(q=0.5)

Year           2015.0
IMDb              7.3
Netflix           0.0
Hulu              0.0
Prime Video       0.0
Disney+           0.0
type              1.0
new_column        7.0
Name: 0.5, dtype: float64

In [78]:
df.var()

Year           121.112570
IMDb             1.281559
Netflix          0.225750
Hulu             0.214920
Prime Video      0.236143
Disney+          0.031056
type             0.000000
new_column       0.000000
dtype: float64

In [79]:
df.std()

Year           11.005116
IMDb            1.132060
Netflix         0.475131
Hulu            0.463594
Prime Video     0.485946
Disney+         0.176228
type            0.000000
new_column      0.000000
dtype: float64

In [81]:
# Counts only non-empty values
df.count()

Title                5611
Year                 5611
Age                  3165
IMDb                 4450
Rotten Tomatoes      1011
Netflix              5611
Hulu                 5611
Prime Video          5611
Disney+              5611
type                 5611
new_column           5611
duplicated_column    5611
todolist             5611
dtype: int64

In [82]:
df.describe()

Year         IMDb      Netflix         Hulu  Prime Video  \
count  5611.000000  4450.000000  5611.000000  5611.000000  5611.000000   
mean   2011.021030     7.113258     0.344145     0.312600     0.382107   
std      11.005116     1.132060     0.475131     0.463594     0.485946   
min    1901.000000     1.000000     0.000000     0.000000     0.000000   
25%    2010.000000     6.600000     0.000000     0.000000     0.000000   
50%    2015.000000     7.300000     0.000000     0.000000     0.000000   
75%    2017.000000     7.900000     1.000000     1.000000     1.000000   
max    2020.000000     9.600000     1.000000     1.000000     1.000000   

           Disney+    type  new_column  
count  5611.000000  5611.0      5611.0  
mean      0.032080     1.0         7.0  
std       0.176228     0.0         0.0  
min       0.000000     1.0         7.0  
25%       0.000000     1.0         7.0  
50%       0.000000     1.0         7.0  
75%       0.000000     1.0         7.0  
max       1.000000     1.0         7.0

In [84]:
df.describe().T

count         mean        std     min     25%     50%     75%  \
Year         5611.0  2011.021030  11.005116  1901.0  2010.0  2015.0  2017.0   
IMDb         4450.0     7.113258   1.132060     1.0     6.6     7.3     7.9   
Netflix      5611.0     0.344145   0.475131     0.0     0.0     0.0     1.0   
Hulu         5611.0     0.312600   0.463594     0.0     0.0     0.0     1.0   
Prime Video  5611.0     0.382107   0.485946     0.0     0.0     0.0     1.0   
Disney+      5611.0     0.032080   0.176228     0.0     0.0     0.0     0.0   
type         5611.0     1.000000   0.000000     1.0     1.0     1.0     1.0   
new_column   5611.0     7.000000   0.000000     7.0     7.0     7.0     7.0   

                max  
Year         2020.0  
IMDb            9.6  
Netflix         1.0  
Hulu            1.0  
Prime Video     1.0  
Disney+         1.0  
type            1.0  
new_column      7.0

In [86]:
df.describe(include='all').T

# Top = first row on the method mode()
# and freq is the number of times it appears

count unique      top  freq        mean        std  \
Title                5611   5564  Kingdom     3         NaN        NaN   
Year               5611.0    NaN      NaN   NaN  2011.02103  11.005116   
Age                  3165      5      16+  1018         NaN        NaN   
IMDb               4450.0    NaN      NaN   NaN    7.113258    1.13206   
Rotten Tomatoes      1011     88     100%   109         NaN        NaN   
Netflix            5611.0    NaN      NaN   NaN    0.344145   0.475131   
Hulu               5611.0    NaN      NaN   NaN      0.3126   0.463594   
Prime Video        5611.0    NaN      NaN   NaN    0.382107   0.485946   
Disney+            5611.0    NaN      NaN   NaN     0.03208   0.176228   
type               5611.0    NaN      NaN   NaN         1.0        0.0   
new_column         5611.0    NaN      NaN   NaN         7.0        0.0   
duplicated_column    5611   5564  Kingdom     3         NaN        NaN   
todolist             5611      2     skip  5601         NaN        NaN   

                      min     25%     50%     75%     max  
Title                 NaN     NaN     NaN     NaN     NaN  
Year               1901.0  2010.0  2015.0  2017.0  2020.0  
Age                   NaN     NaN     NaN     NaN     NaN  
IMDb                  1.0     6.6     7.3     7.9     9.6  
Rotten Tomatoes       NaN     NaN     NaN     NaN     NaN  
Netflix               0.0     0.0     0.0     1.0     1.0  
Hulu                  0.0     0.0     0.0     1.0     1.0  
Prime Video           0.0     0.0     0.0     1.0     1.0  
Disney+               0.0     0.0     0.0     0.0     1.0  
type                  1.0     1.0     1.0     1.0     1.0  
new_column            7.0     7.0     7.0     7.0     7.0  
duplicated_column     NaN     NaN     NaN     NaN     NaN  
todolist              NaN     NaN     NaN     NaN     NaN

# Calculations

In [87]:
df._get_numeric_data()

Year  IMDb  Netflix  Hulu  Prime Video  Disney+  type  new_column
0     2008   9.5        1     0            0        0     1           7
1     2016   8.8        1     0            0        0     1           7
2     2017   8.4        1     0            0        0     1           7
3     2010   9.1        1     0            0        0     1           7
4     2015   8.7        1     0            0        0     1           7
...    ...   ...      ...   ...          ...      ...   ...         ...
5606  2018   NaN        0     0            0        1     1           7
5607  2017   NaN        0     0            0        1     1           7
5608  2018   NaN        0     0            0        1     1           7
5609  2017   NaN        0     0            0        1     1           7
5610  2016   NaN        0     0            0        1     1           7

[5611 rows x 8 columns]

In [89]:
df.sum()

Title                Breaking BadStranger ThingsMoney HeistSherlock...
Year                                                          11283839
IMDb                                                           31654.0
Netflix                                                           1931
Hulu                                                              1754
Prime Video                                                       2144
Disney+                                                            180
type                                                              5611
new_column                                                       39277
duplicated_column    Breaking BadStranger ThingsMoney HeistSherlock...
todolist             to watchskipskipskipskipskipskipskipskipskipsk...
dtype: object

In [90]:
df.sum(axis=1)

0       2026.5
1       2033.8
2       2034.4
3       2028.1
4       2032.7
         ...  
5606    2027.0
5607    2026.0
5608    2027.0
5609    2026.0
5610    2025.0
Length: 5611, dtype: float64

In [91]:
df[['Netflix', 'Hulu']].sum(axis=1)

0       1
1       1
2       1
3       1
4       1
       ..
5606    0
5607    0
5608    0
5609    0
5610    0
Length: 5611, dtype: int64

In [92]:
df.Hulu+df.Netflix

0       1
1       1
2       1
3       1
4       1
       ..
5606    0
5607    0
5608    0
5609    0
5610    0
Length: 5611, dtype: int64

In [94]:
df.IMDb+df.Netflix

0       10.5
1        9.8
2        9.4
3       10.1
4        9.7
        ... 
5606     NaN
5607     NaN
5608     NaN
5609     NaN
5610     NaN
Length: 5611, dtype: float64

In [97]:
df.Netflix.add(df.IMDb,fill_value=0)

0       10.5
1        9.8
2        9.4
3       10.1
4        9.7
        ... 
5606     0.0
5607     0.0
5608     0.0
5609     0.0
5610     0.0
Length: 5611, dtype: float64

In [99]:
df.Netflix.sub(df['Disney+'])

0       1
1       1
2       1
3       1
4       1
       ..
5606   -1
5607   -1
5608   -1
5609   -1
5610   -1
Length: 5611, dtype: int64

In [100]:
df._get_numeric_data()

Year  IMDb  Netflix  Hulu  Prime Video  Disney+  type  new_column
0     2008   9.5        1     0            0        0     1           7
1     2016   8.8        1     0            0        0     1           7
2     2017   8.4        1     0            0        0     1           7
3     2010   9.1        1     0            0        0     1           7
4     2015   8.7        1     0            0        0     1           7
...    ...   ...      ...   ...          ...      ...   ...         ...
5606  2018   NaN        0     0            0        1     1           7
5607  2017   NaN        0     0            0        1     1           7
5608  2018   NaN        0     0            0        1     1           7
5609  2017   NaN        0     0            0        1     1           7
5610  2016   NaN        0     0            0        1     1           7

[5611 rows x 8 columns]

In [102]:
df._get_numeric_data().diff()

Year  IMDb  Netflix  Hulu  Prime Video  Disney+  type  new_column
0      NaN   NaN      NaN   NaN          NaN      NaN   NaN         NaN
1      8.0  -0.7      0.0   0.0          0.0      0.0   0.0         0.0
2      1.0  -0.4      0.0   0.0          0.0      0.0   0.0         0.0
3     -7.0   0.7      0.0   0.0          0.0      0.0   0.0         0.0
4      5.0  -0.4      0.0   0.0          0.0      0.0   0.0         0.0
...    ...   ...      ...   ...          ...      ...   ...         ...
5606   5.0   NaN      0.0   0.0          0.0      0.0   0.0         0.0
5607  -1.0   NaN      0.0   0.0          0.0      0.0   0.0         0.0
5608   1.0   NaN      0.0   0.0          0.0      0.0   0.0         0.0
5609  -1.0   NaN      0.0   0.0          0.0      0.0   0.0         0.0
5610  -1.0   NaN      0.0   0.0          0.0      0.0   0.0         0.0

[5611 rows x 8 columns]

In [103]:
df.Hulu*df.Netflix

0       0
1       0
2       0
3       0
4       0
       ..
5606    0
5607    0
5608    0
5609    0
5610    0
Length: 5611, dtype: int64

In [105]:
df.Hulu.mul(df.Netflix)

0       0
1       0
2       0
3       0
4       0
       ..
5606    0
5607    0
5608    0
5609    0
5610    0
Length: 5611, dtype: int64

In [106]:
df.Hulu/df["Prime Video"]

0      NaN
1      NaN
2      NaN
3      NaN
4      NaN
        ..
5606   NaN
5607   NaN
5608   NaN
5609   NaN
5610   NaN
Length: 5611, dtype: float64

In [108]:
df.Hulu.div(df["Prime Video"])

0      NaN
1      NaN
2      NaN
3      NaN
4      NaN
        ..
5606   NaN
5607   NaN
5608   NaN
5609   NaN
5610   NaN
Length: 5611, dtype: float64

In [110]:
df.Hulu.rdiv(df["Prime Video"])

0      NaN
1      NaN
2      NaN
3      NaN
4      NaN
        ..
5606   NaN
5607   NaN
5608   NaN
5609   NaN
5610   NaN
Length: 5611, dtype: float64

# Missing data

In [111]:
df.isna()

Title   Year    Age   IMDb  Rotten Tomatoes  Netflix   Hulu  \
0     False  False  False  False            False    False  False   
1     False  False  False  False            False    False  False   
2     False  False  False  False            False    False  False   
3     False  False  False  False            False    False  False   
4     False  False  False  False            False    False  False   
...     ...    ...    ...    ...              ...      ...    ...   
5606  False  False   True   True             True    False  False   
5607  False  False   True   True             True    False  False   
5608  False  False   True   True             True    False  False   
5609  False  False   True   True             True    False  False   
5610  False  False   True   True             True    False  False   

      Prime Video  Disney+   type  new_column  duplicated_column  todolist  
0           False    False  False       False              False     False  
1           False    False  False       False              False     False  
2           False    False  False       False              False     False  
3           False    False  False       False              False     False  
4           False    False  False       False              False     False  
...           ...      ...    ...         ...                ...       ...  
5606        False    False  False       False              False     False  
5607        False    False  False       False              False     False  
5608        False    False  False       False              False     False  
5609        False    False  False       False              False     False  
5610        False    False  False       False              False     False  

[5611 rows x 13 columns]

In [112]:
df.isna().sum()

Title                   0
Year                    0
Age                  2446
IMDb                 1161
Rotten Tomatoes      4600
Netflix                 0
Hulu                    0
Prime Video             0
Disney+                 0
type                    0
new_column              0
duplicated_column       0
todolist                0
dtype: int64

## Sort values

In [114]:
df.sort_values(by="Age")

Title  Year  Age  IMDb Rotten Tomatoes  \
4024                      Dark Knight  2000  13+   5.3             NaN   
1116                           Powder  1995  13+   6.6             50%   
1514                      The In-Laws  2003  13+   5.7             34%   
1123               Behind Enemy Lines  2001  13+   6.4             37%   
2318                Darker than Black  2007  16+   7.7             NaN   
...                               ...   ...  ...   ...             ...   
5606  Tut's Treasures: Hidden Secrets  2018  NaN   NaN             NaN   
5607                 Paradise Islands  2017  NaN   NaN             NaN   
5608                      Wild Russia  2018  NaN   NaN             NaN   
5609                      Love & Vets  2017  NaN   NaN             NaN   
5610         United States of Animals  2016  NaN   NaN             NaN   

      Netflix  Hulu  Prime Video  Disney+  type  new_column  \
4024        0     0            1        0     1           7   
1116        1     0            0        0     1           7   
1514        1     0            0        0     1           7   
1123        1     0            0        0     1           7   
2318        0     1            1        0     1           7   
...       ...   ...          ...      ...   ...         ...   
5606        0     0            0        1     1           7   
5607        0     0            0        1     1           7   
5608        0     0            0        1     1           7   
5609        0     0            0        1     1           7   
5610        0     0            0        1     1           7   

                    duplicated_column todolist  
4024                      Dark Knight     skip  
1116                           Powder     skip  
1514                      The In-Laws     skip  
1123               Behind Enemy Lines     skip  
2318                Darker than Black     skip  
...                               ...      ...  
5606  Tut's Treasures: Hidden Secrets     skip  
5607                 Paradise Islands     skip  
5608                      Wild Russia     skip  
5609                      Love & Vets     skip  
5610         United States of Animals     skip  

[5611 rows x 13 columns]

In [116]:
df.sort_values(by="Year")

Title  Year  Age  IMDb Rotten Tomatoes  \
4679  Gods & Monsters with Tony Robinson  1901  NaN   NaN             NaN   
4652             Space: The New Frontier  1901  NaN   NaN             NaN   
4701             History of Westinghouse  1904  NaN   NaN             NaN   
1854                     Born To Explore  1914  NaN   NaN             NaN   
3328         The Little Rascals Classics  1931  NaN   NaN             NaN   
...                                  ...   ...  ...   ...             ...   
216                                Cheer  2020  16+   8.2             96%   
2796                           Plunderer  2020  16+   6.2             NaN   
402                            Gentefied  2020  16+   7.3             88%   
597                                Freud  2020  18+   6.6             50%   
2001                                DAVE  2020  18+   8.4             62%   

      Netflix  Hulu  Prime Video  Disney+  type  new_column  \
4679        0     0            1        0     1           7   
4652        0     0            1        0     1           7   
4701        0     0            1        0     1           7   
1854        1     0            0        0     1           7   
3328        0     1            0        0     1           7   
...       ...   ...          ...      ...   ...         ...   
216         1     0            0        0     1           7   
2796        0     1            0        0     1           7   
402         1     0            0        0     1           7   
597         1     0            0        0     1           7   
2001        0     1            0        0     1           7   

                       duplicated_column todolist  
4679  Gods & Monsters with Tony Robinson     skip  
4652             Space: The New Frontier     skip  
4701             History of Westinghouse     skip  
1854                     Born To Explore     skip  
3328         The Little Rascals Classics     skip  
...                                  ...      ...  
216                                Cheer     skip  
2796                           Plunderer     skip  
402                            Gentefied     skip  
597                                Freud     skip  
2001                                DAVE     skip  

[5611 rows x 13 columns]

In [118]:
df.sort_values(by=["Year","Title"])

Title  Year  Age  IMDb Rotten Tomatoes  \
4679  Gods & Monsters with Tony Robinson  1901  NaN   NaN             NaN   
4652             Space: The New Frontier  1901  NaN   NaN             NaN   
4701             History of Westinghouse  1904  NaN   NaN             NaN   
1854                     Born To Explore  1914  NaN   NaN             NaN   
3328         The Little Rascals Classics  1931  NaN   NaN             NaN   
...                                  ...   ...  ...   ...             ...   
668                          White Lines  2020  18+   6.2             58%   
2878       Who Wants to Be a Millionaire  2020  16+   6.1             NaN   
727           Win the Wilderness: Alaska  2020  NaN   7.1             NaN   
3646                        ZeroZeroZero  2020  18+   8.2             80%   
2026       Zoey's Extraordinary Playlist  2020  16+   8.1             74%   

      Netflix  Hulu  Prime Video  Disney+  type  new_column  \
4679        0     0            1        0     1           7   
4652        0     0            1        0     1           7   
4701        0     0            1        0     1           7   
1854        1     0            0        0     1           7   
3328        0     1            0        0     1           7   
...       ...   ...          ...      ...   ...         ...   
668         1     0            0        0     1           7   
2878        0     1            0        0     1           7   
727         1     0            0        0     1           7   
3646        0     0            1        0     1           7   
2026        0     1            0        0     1           7   

                       duplicated_column todolist  
4679  Gods & Monsters with Tony Robinson     skip  
4652             Space: The New Frontier     skip  
4701             History of Westinghouse     skip  
1854                     Born To Explore     skip  
3328         The Little Rascals Classics     skip  
...                                  ...      ...  
668                          White Lines     skip  
2878       Who Wants to Be a Millionaire     skip  
727           Win the Wilderness: Alaska     skip  
3646                        ZeroZeroZero     skip  
2026       Zoey's Extraordinary Playlist     skip  

[5611 rows x 13 columns]

In [120]:
df.sort_values(by=df.columns[:3].tolist())

Title  Year  Age  IMDb Rotten Tomatoes  Netflix  Hulu  \
4638   #MeToo, Now What?  2018  NaN   6.6             NaN        0     0   
3409    #ThatsHarassment  2018  NaN   NaN             NaN        0     1   
439             #blackAF  2020  18+   6.6             NaN        1     0   
696   (The Hook Up Plan)  2018  16+   7.4             NaN        1     0   
3061               .hack  2002   7+   6.9             18%        0     1   
...                  ...   ...  ...   ...             ...      ...   ...   
1740               Тобот  2010  NaN   6.2             NaN        1     0   
2975         恋と選挙とチョコレート  2012  18+   7.2             NaN        0     1   
4336             那年花開月正圓  2017   7+   7.8             NaN        0     0   
1789                 阳关道  2018  NaN   3.1             NaN        1     0   
2850    頭文字D First Stage  1998   7+   8.4             NaN        0     1   

      Prime Video  Disney+  type  new_column   duplicated_column todolist  
4638            1        0     1           7   #MeToo, Now What?     skip  
3409            1        0     1           7    #ThatsHarassment     skip  
439             0        0     1           7            #blackAF     skip  
696             0        0     1           7  (The Hook Up Plan)     skip  
3061            0        0     1           7               .hack     skip  
...           ...      ...   ...         ...                 ...      ...  
1740            1        0     1           7               Тобот     skip  
2975            0        0     1           7         恋と選挙とチョコレート     skip  
4336            1        0     1           7             那年花開月正圓     skip  
1789            0        0     1           7                 阳关道     skip  
2850            0        0     1           7    頭文字D First Stage     skip  

[5611 rows x 13 columns]

In [122]:
df.sort_values(by="Year", ascending=False)

Title  Year  Age  IMDb Rotten Tomatoes  \
2537                         Duncanville  2020  16+   5.6             89%   
671               Coronavirus, Explained  2020  NaN   7.8             NaN   
2037             Little Fires Everywhere  2020  18+   7.7             77%   
1064                     The Ghost Bride  2020  18+   5.8             NaN   
155                         The Stranger  2020  16+   7.3             89%   
...                                  ...   ...  ...   ...             ...   
3328         The Little Rascals Classics  1931  NaN   NaN             NaN   
1854                     Born To Explore  1914  NaN   NaN             NaN   
4701             History of Westinghouse  1904  NaN   NaN             NaN   
4652             Space: The New Frontier  1901  NaN   NaN             NaN   
4679  Gods & Monsters with Tony Robinson  1901  NaN   NaN             NaN   

      Netflix  Hulu  Prime Video  Disney+  type  new_column  \
2537        0     1            0        0     1           7   
671         1     0            0        0     1           7   
2037        0     1            0        0     1           7   
1064        1     0            0        0     1           7   
155         1     0            0        0     1           7   
...       ...   ...          ...      ...   ...         ...   
3328        0     1            0        0     1           7   
1854        1     0            0        0     1           7   
4701        0     0            1        0     1           7   
4652        0     0            1        0     1           7   
4679        0     0            1        0     1           7   

                       duplicated_column todolist  
2537                         Duncanville     skip  
671               Coronavirus, Explained     skip  
2037             Little Fires Everywhere     skip  
1064                     The Ghost Bride     skip  
155                         The Stranger     skip  
...                                  ...      ...  
3328         The Little Rascals Classics     skip  
1854                     Born To Explore     skip  
4701             History of Westinghouse     skip  
4652             Space: The New Frontier     skip  
4679  Gods & Monsters with Tony Robinson     skip  

[5611 rows x 13 columns]

In [124]:
df.sort_values(by="IMDb", ascending=False, na_position='first')

Title  Year  Age  IMDb Rotten Tomatoes  \
871                    Find Yourself  2020  NaN   NaN             NaN   
967              A Love So Beautiful  2017  NaN   NaN             NaN   
1136                 To Be Continued  2015  all   NaN             NaN   
1194           Legend of Chu and Han  2012  NaN   NaN             NaN   
1334              Total Drama Action  2009   7+   NaN             NaN   
...                              ...   ...  ...   ...             ...   
3104               Toddlers & Tiaras  2009   7+   1.8             NaN   
1818                Game Winning Hit  2009  NaN   1.8             NaN   
3145                    The Proposal  2018  NaN   1.8             NaN   
2999  A Little Late with Lilly Singh  2019  16+   1.7             90%   
1807                     Be With You  2015  NaN   1.0             NaN   

      Netflix  Hulu  Prime Video  Disney+  type  new_column  \
871         1     0            0        0     1           7   
967         1     0            0        0     1           7   
1136        1     0            0        0     1           7   
1194        1     0            0        0     1           7   
1334        1     0            0        0     1           7   
...       ...   ...          ...      ...   ...         ...   
3104        0     1            1        0     1           7   
1818        1     0            0        0     1           7   
3145        0     1            0        0     1           7   
2999        0     1            0        0     1           7   
1807        1     0            0        0     1           7   

                   duplicated_column todolist  
871                    Find Yourself     skip  
967              A Love So Beautiful     skip  
1136                 To Be Continued     skip  
1194           Legend of Chu and Han     skip  
1334              Total Drama Action     skip  
...                              ...      ...  
3104               Toddlers & Tiaras     skip  
1818                Game Winning Hit     skip  
3145                    The Proposal     skip  
2999  A Little Late with Lilly Singh     skip  
1807                     Be With You     skip  

[5611 rows x 13 columns]

In [126]:
df.sort_values(by="IMDb", ascending=False, na_position='first', ignore_index=True)

Title  Year  Age  IMDb Rotten Tomatoes  \
0                      Find Yourself  2020  NaN   NaN             NaN   
1                A Love So Beautiful  2017  NaN   NaN             NaN   
2                    To Be Continued  2015  all   NaN             NaN   
3              Legend of Chu and Han  2012  NaN   NaN             NaN   
4                 Total Drama Action  2009   7+   NaN             NaN   
...                              ...   ...  ...   ...             ...   
5606               Toddlers & Tiaras  2009   7+   1.8             NaN   
5607                Game Winning Hit  2009  NaN   1.8             NaN   
5608                    The Proposal  2018  NaN   1.8             NaN   
5609  A Little Late with Lilly Singh  2019  16+   1.7             90%   
5610                     Be With You  2015  NaN   1.0             NaN   

      Netflix  Hulu  Prime Video  Disney+  type  new_column  \
0           1     0            0        0     1           7   
1           1     0            0        0     1           7   
2           1     0            0        0     1           7   
3           1     0            0        0     1           7   
4           1     0            0        0     1           7   
...       ...   ...          ...      ...   ...         ...   
5606        0     1            1        0     1           7   
5607        1     0            0        0     1           7   
5608        0     1            0        0     1           7   
5609        0     1            0        0     1           7   
5610        1     0            0        0     1           7   

                   duplicated_column todolist  
0                      Find Yourself     skip  
1                A Love So Beautiful     skip  
2                    To Be Continued     skip  
3              Legend of Chu and Han     skip  
4                 Total Drama Action     skip  
...                              ...      ...  
5606               Toddlers & Tiaras     skip  
5607                Game Winning Hit     skip  
5608                    The Proposal     skip  
5609  A Little Late with Lilly Singh     skip  
5610                     Be With You     skip  

[5611 rows x 13 columns]

In [127]:
df.sort_values(by="IMDb", ascending=False, na_position='first', ignore_index=True, inplace=True)

In [128]:
df.head()

Title  Year  Age  IMDb Rotten Tomatoes  Netflix  Hulu  \
0          Find Yourself  2020  NaN   NaN             NaN        1     0   
1    A Love So Beautiful  2017  NaN   NaN             NaN        1     0   
2        To Be Continued  2015  all   NaN             NaN        1     0   
3  Legend of Chu and Han  2012  NaN   NaN             NaN        1     0   
4     Total Drama Action  2009   7+   NaN             NaN        1     0   

   Prime Video  Disney+  type  new_column      duplicated_column todolist  
0            0        0     1           7          Find Yourself     skip  
1            0        0     1           7    A Love So Beautiful     skip  
2            0        0     1           7        To Be Continued     skip  
3            0        0     1           7  Legend of Chu and Han     skip  
4            0        0     1           7     Total Drama Action     skip

In [131]:
df.sort_values(by="Title", ascending=False, na_position='first', inplace=True)

In [132]:
df.head()

Title  Year  Age  IMDb Rotten Tomatoes  Netflix  Hulu  \
1506  頭文字D First Stage  1998   7+   8.4             NaN        0     1   
5577               阳关道  2018  NaN   3.1             NaN        1     0   
2476           那年花開月正圓  2017   7+   7.8             NaN        0     0   
3543       恋と選挙とチョコレート  2012  18+   7.2             NaN        0     1   
4856             Тобот  2010  NaN   6.2             NaN        1     0   

      Prime Video  Disney+  type  new_column duplicated_column todolist  
1506            0        0     1           7  頭文字D First Stage     skip  
5577            0        0     1           7               阳关道     skip  
2476            1        0     1           7           那年花開月正圓     skip  
3543            0        0     1           7       恋と選挙とチョコレート     skip  
4856            1        0     1           7             Тобот     skip

## Sort index

In [133]:
df.sort_index()

Title  Year  Age  IMDb Rotten Tomatoes  \
0                      Find Yourself  2020  NaN   NaN             NaN   
1                A Love So Beautiful  2017  NaN   NaN             NaN   
2                    To Be Continued  2015  all   NaN             NaN   
3              Legend of Chu and Han  2012  NaN   NaN             NaN   
4                 Total Drama Action  2009   7+   NaN             NaN   
...                              ...   ...  ...   ...             ...   
5606               Toddlers & Tiaras  2009   7+   1.8             NaN   
5607                Game Winning Hit  2009  NaN   1.8             NaN   
5608                    The Proposal  2018  NaN   1.8             NaN   
5609  A Little Late with Lilly Singh  2019  16+   1.7             90%   
5610                     Be With You  2015  NaN   1.0             NaN   

      Netflix  Hulu  Prime Video  Disney+  type  new_column  \
0           1     0            0        0     1           7   
1           1     0            0        0     1           7   
2           1     0            0        0     1           7   
3           1     0            0        0     1           7   
4           1     0            0        0     1           7   
...       ...   ...          ...      ...   ...         ...   
5606        0     1            1        0     1           7   
5607        1     0            0        0     1           7   
5608        0     1            0        0     1           7   
5609        0     1            0        0     1           7   
5610        1     0            0        0     1           7   

                   duplicated_column todolist  
0                      Find Yourself     skip  
1                A Love So Beautiful     skip  
2                    To Be Continued     skip  
3              Legend of Chu and Han     skip  
4                 Total Drama Action     skip  
...                              ...      ...  
5606               Toddlers & Tiaras     skip  
5607                Game Winning Hit     skip  
5608                    The Proposal     skip  
5609  A Little Late with Lilly Singh     skip  
5610                     Be With You     skip  

[5611 rows x 13 columns]

# Drop columns

In [134]:
df.drop(columns='duplicated_column')

Title  Year  Age  IMDb Rotten Tomatoes  Netflix  Hulu  \
1506    頭文字D First Stage  1998   7+   8.4             NaN        0     1   
5577                 阳关道  2018  NaN   3.1             NaN        1     0   
2476             那年花開月正圓  2017   7+   7.8             NaN        0     0   
3543         恋と選挙とチョコレート  2012  18+   7.2             NaN        0     1   
4856               Тобот  2010  NaN   6.2             NaN        1     0   
...                  ...   ...  ...   ...             ...      ...   ...   
4122               .hack  2002   7+   6.9             18%        0     1   
3200  (The Hook Up Plan)  2018  16+   7.4             NaN        1     0   
4492            #blackAF  2020  18+   6.6             NaN        1     0   
199     #ThatsHarassment  2018  NaN   NaN             NaN        0     1   
4420   #MeToo, Now What?  2018  NaN   6.6             NaN        0     0   

      Prime Video  Disney+  type  new_column todolist  
1506            0        0     1           7     skip  
5577            0        0     1           7     skip  
2476            1        0     1           7     skip  
3543            0        0     1           7     skip  
4856            1        0     1           7     skip  
...           ...      ...   ...         ...      ...  
4122            0        0     1           7     skip  
3200            0        0     1           7     skip  
4492            0        0     1           7     skip  
199             1        0     1           7     skip  
4420            1        0     1           7     skip  

[5611 rows x 12 columns]

In [135]:
df.drop('duplicated_column', axis=1)

Title  Year  Age  IMDb Rotten Tomatoes  Netflix  Hulu  \
1506    頭文字D First Stage  1998   7+   8.4             NaN        0     1   
5577                 阳关道  2018  NaN   3.1             NaN        1     0   
2476             那年花開月正圓  2017   7+   7.8             NaN        0     0   
3543         恋と選挙とチョコレート  2012  18+   7.2             NaN        0     1   
4856               Тобот  2010  NaN   6.2             NaN        1     0   
...                  ...   ...  ...   ...             ...      ...   ...   
4122               .hack  2002   7+   6.9             18%        0     1   
3200  (The Hook Up Plan)  2018  16+   7.4             NaN        1     0   
4492            #blackAF  2020  18+   6.6             NaN        1     0   
199     #ThatsHarassment  2018  NaN   NaN             NaN        0     1   
4420   #MeToo, Now What?  2018  NaN   6.6             NaN        0     0   

      Prime Video  Disney+  type  new_column todolist  
1506            0        0     1           7     skip  
5577            0        0     1           7     skip  
2476            1        0     1           7     skip  
3543            0        0     1           7     skip  
4856            1        0     1           7     skip  
...           ...      ...   ...         ...      ...  
4122            0        0     1           7     skip  
3200            0        0     1           7     skip  
4492            0        0     1           7     skip  
199             1        0     1           7     skip  
4420            1        0     1           7     skip  

[5611 rows x 12 columns]

In [136]:
df1=df.drop(['new_column', 'todolist','duplicated_column'],axis=1)

In [137]:
df1.head()

Title  Year  Age  IMDb Rotten Tomatoes  Netflix  Hulu  \
1506  頭文字D First Stage  1998   7+   8.4             NaN        0     1   
5577               阳关道  2018  NaN   3.1             NaN        1     0   
2476           那年花開月正圓  2017   7+   7.8             NaN        0     0   
3543       恋と選挙とチョコレート  2012  18+   7.2             NaN        0     1   
4856             Тобот  2010  NaN   6.2             NaN        1     0   

      Prime Video  Disney+  type  
1506            0        0     1  
5577            0        0     1  
2476            1        0     1  
3543            0        0     1  
4856            1        0     1

#  Rows manipulations

## Change datatype

In [138]:
df.Year.astype(str)

1506    1998
5577    2018
2476    2017
3543    2012
4856    2010
        ... 
4122    2002
3200    2018
4492    2020
199     2018
4420    2018
Name: Year, Length: 5611, dtype: object

## Set/reset index

In [140]:
df.Title+', '+df.Year.astype(str)

1506      頭文字D First Stage, 1998
5577                   阳关道, 2018
2476               那年花開月正圓, 2017
3543           恋と選挙とチョコレート, 2012
4856                 Тобот, 2010
                  ...           
4122                 .hack, 2002
3200    (The Hook Up Plan), 2018
4492              #blackAF, 2020
199       #ThatsHarassment, 2018
4420     #MeToo, Now What?, 2018
Length: 5611, dtype: object

In [142]:
(df.Title+', '+df.Year.astype(str)).nunique()

5611

In [144]:
df.shape[0]

5611

We can see that the number of unique combinations of title and year give us a unique identifier of a TV series.
Let's set them as index.

In [146]:
df.set_index(["Title","Year"])

Age  IMDb Rotten Tomatoes  Netflix  Hulu  \
Title              Year                                             
頭文字D First Stage   1998   7+   8.4             NaN        0     1   
阳关道                2018  NaN   3.1             NaN        1     0   
那年花開月正圓            2017   7+   7.8             NaN        0     0   
恋と選挙とチョコレート        2012  18+   7.2             NaN        0     1   
Тобот              2010  NaN   6.2             NaN        1     0   
...                      ...   ...             ...      ...   ...   
.hack              2002   7+   6.9             18%        0     1   
(The Hook Up Plan) 2018  16+   7.4             NaN        1     0   
#blackAF           2020  18+   6.6             NaN        1     0   
#ThatsHarassment   2018  NaN   NaN             NaN        0     1   
#MeToo, Now What?  2018  NaN   6.6             NaN        0     0   

                         Prime Video  Disney+  type  new_column  \
Title              Year                                           
頭文字D First Stage   1998            0        0     1           7   
阳关道                2018            0        0     1           7   
那年花開月正圓            2017            1        0     1           7   
恋と選挙とチョコレート        2012            0        0     1           7   
Тобот              2010            1        0     1           7   
...                              ...      ...   ...         ...   
.hack              2002            0        0     1           7   
(The Hook Up Plan) 2018            0        0     1           7   
#blackAF           2020            0        0     1           7   
#ThatsHarassment   2018            1        0     1           7   
#MeToo, Now What?  2018            1        0     1           7   

                          duplicated_column todolist  
Title              Year                               
頭文字D First Stage   1998    頭文字D First Stage     skip  
阳关道                2018                 阳关道     skip  
那年花開月正圓            2017             那年花開月正圓     skip  
恋と選挙とチョコレート        2012         恋と選挙とチョコレート     skip  
Тобот              2010               Тобот     skip  
...                                     ...      ...  
.hack              2002               .hack     skip  
(The Hook Up Plan) 2018  (The Hook Up Plan)     skip  
#blackAF           2020            #blackAF     skip  
#ThatsHarassment   2018    #ThatsHarassment     skip  
#MeToo, Now What?  2018   #MeToo, Now What?     skip  

[5611 rows x 11 columns]

In [147]:
df=df.set_index(["Title","Year"])

In [148]:
df.head()

Age  IMDb Rotten Tomatoes  Netflix  Hulu  Prime Video  \
Title            Year                                                          
頭文字D First Stage 1998   7+   8.4             NaN        0     1            0   
阳关道              2018  NaN   3.1             NaN        1     0            0   
那年花開月正圓          2017   7+   7.8             NaN        0     0            1   
恋と選挙とチョコレート      2012  18+   7.2             NaN        0     1            0   
Тобот            2010  NaN   6.2             NaN        1     0            1   

                       Disney+  type  new_column duplicated_column todolist  
Title            Year                                                        
頭文字D First Stage 1998        0     1           7  頭文字D First Stage     skip  
阳关道              2018        0     1           7               阳关道     skip  
那年花開月正圓          2017        0     1           7           那年花開月正圓     skip  
恋と選挙とチョコレート      2012        0     1           7       恋と選挙とチョコレート     skip  
Тобот            2010        0     1           7             Тобот     skip

In [ ]:
df.index

In [151]:
df.loc[("Тобот")]

Age  IMDb Rotten Tomatoes  Netflix  Hulu  Prime Video  Disney+  type  \
Year                                                                         
2010  NaN   6.2             NaN        1     0            1        0     1   

      new_column duplicated_column todolist  
Year                                         
2010           7             Тобот     skip

In [153]:
df.loc["Undercover"]

Age  IMDb Rotten Tomatoes  Netflix  Hulu  Prime Video  Disney+  type  \
Year                                                                         
2011   7+   8.9             NaN        0     0            1        0     1   
2016  16+   6.8             75%        0     0            1        0     1   
2019  18+   7.9             NaN        1     0            0        0     1   

      new_column duplicated_column todolist  
Year                                         
2011           7        Undercover     skip  
2016           7        Undercover     skip  
2019           7        Undercover     skip

In [156]:
df.head()

Age  IMDb Rotten Tomatoes  Netflix  Hulu  Prime Video  \
Title            Year                                                          
頭文字D First Stage 1998   7+   8.4             NaN        0     1            0   
阳关道              2018  NaN   3.1             NaN        1     0            0   
那年花開月正圓          2017   7+   7.8             NaN        0     0            1   
恋と選挙とチョコレート      2012  18+   7.2             NaN        0     1            0   
Тобот            2010  NaN   6.2             NaN        1     0            1   

                       Disney+  type  new_column duplicated_column todolist  
Title            Year                                                        
頭文字D First Stage 1998        0     1           7  頭文字D First Stage     skip  
阳关道              2018        0     1           7               阳关道     skip  
那年花開月正圓          2017        0     1           7           那年花開月正圓     skip  
恋と選挙とチョコレート      2012        0     1           7       恋と選挙とチョコレート     skip  
Тобот            2010        0     1           7             Тобот     skip

In [157]:
df.reset_index()

Title  Year  Age  IMDb Rotten Tomatoes  Netflix  Hulu  \
0       頭文字D First Stage  1998   7+   8.4             NaN        0     1   
1                    阳关道  2018  NaN   3.1             NaN        1     0   
2                那年花開月正圓  2017   7+   7.8             NaN        0     0   
3            恋と選挙とチョコレート  2012  18+   7.2             NaN        0     1   
4                  Тобот  2010  NaN   6.2             NaN        1     0   
...                  ...   ...  ...   ...             ...      ...   ...   
5606               .hack  2002   7+   6.9             18%        0     1   
5607  (The Hook Up Plan)  2018  16+   7.4             NaN        1     0   
5608            #blackAF  2020  18+   6.6             NaN        1     0   
5609    #ThatsHarassment  2018  NaN   NaN             NaN        0     1   
5610   #MeToo, Now What?  2018  NaN   6.6             NaN        0     0   

      Prime Video  Disney+  type  new_column   duplicated_column todolist  
0               0        0     1           7    頭文字D First Stage     skip  
1               0        0     1           7                 阳关道     skip  
2               1        0     1           7             那年花開月正圓     skip  
3               0        0     1           7         恋と選挙とチョコレート     skip  
4               1        0     1           7               Тобот     skip  
...           ...      ...   ...         ...                 ...      ...  
5606            0        0     1           7               .hack     skip  
5607            0        0     1           7  (The Hook Up Plan)     skip  
5608            0        0     1           7            #blackAF     skip  
5609            1        0     1           7    #ThatsHarassment     skip  
5610            1        0     1           7   #MeToo, Now What?     skip  

[5611 rows x 13 columns]

In [158]:
df.reset_index(level=1)

Year  Age  IMDb Rotten Tomatoes  Netflix  Hulu  \
Title                                                                
頭文字D First Stage    1998   7+   8.4             NaN        0     1   
阳关道                 2018  NaN   3.1             NaN        1     0   
那年花開月正圓             2017   7+   7.8             NaN        0     0   
恋と選挙とチョコレート         2012  18+   7.2             NaN        0     1   
Тобот               2010  NaN   6.2             NaN        1     0   
...                  ...  ...   ...             ...      ...   ...   
.hack               2002   7+   6.9             18%        0     1   
(The Hook Up Plan)  2018  16+   7.4             NaN        1     0   
#blackAF            2020  18+   6.6             NaN        1     0   
#ThatsHarassment    2018  NaN   NaN             NaN        0     1   
#MeToo, Now What?   2018  NaN   6.6             NaN        0     0   

                    Prime Video  Disney+  type  new_column  \
Title                                                        
頭文字D First Stage              0        0     1           7   
阳关道                           0        0     1           7   
那年花開月正圓                       1        0     1           7   
恋と選挙とチョコレート                   0        0     1           7   
Тобот                         1        0     1           7   
...                         ...      ...   ...         ...   
.hack                         0        0     1           7   
(The Hook Up Plan)            0        0     1           7   
#blackAF                      0        0     1           7   
#ThatsHarassment              1        0     1           7   
#MeToo, Now What?             1        0     1           7   

                     duplicated_column todolist  
Title                                            
頭文字D First Stage      頭文字D First Stage     skip  
阳关道                                阳关道     skip  
那年花開月正圓                        那年花開月正圓     skip  
恋と選挙とチョコレート                恋と選挙とチョコレート     skip  
Тобот                            Тобот     skip  
...                                ...      ...  
.hack                            .hack     skip  
(The Hook Up Plan)  (The Hook Up Plan)     skip  
#blackAF                      #blackAF     skip  
#ThatsHarassment      #ThatsHarassment     skip  
#MeToo, Now What?    #MeToo, Now What?     skip  

[5611 rows x 12 columns]

In [159]:
df.reset_index(level=1,drop=True)

Age  IMDb Rotten Tomatoes  Netflix  Hulu  Prime Video  \
Title                                                                       
頭文字D First Stage     7+   8.4             NaN        0     1            0   
阳关道                 NaN   3.1             NaN        1     0            0   
那年花開月正圓              7+   7.8             NaN        0     0            1   
恋と選挙とチョコレート         18+   7.2             NaN        0     1            0   
Тобот               NaN   6.2             NaN        1     0            1   
...                 ...   ...             ...      ...   ...          ...   
.hack                7+   6.9             18%        0     1            0   
(The Hook Up Plan)  16+   7.4             NaN        1     0            0   
#blackAF            18+   6.6             NaN        1     0            0   
#ThatsHarassment    NaN   NaN             NaN        0     1            1   
#MeToo, Now What?   NaN   6.6             NaN        0     0            1   

                    Disney+  type  new_column   duplicated_column todolist  
Title                                                                       
頭文字D First Stage          0     1           7    頭文字D First Stage     skip  
阳关道                       0     1           7                 阳关道     skip  
那年花開月正圓                   0     1           7             那年花開月正圓     skip  
恋と選挙とチョコレート               0     1           7         恋と選挙とチョコレート     skip  
Тобот                     0     1           7               Тобот     skip  
...                     ...   ...         ...                 ...      ...  
.hack                     0     1           7               .hack     skip  
(The Hook Up Plan)        0     1           7  (The Hook Up Plan)     skip  
#blackAF                  0     1           7            #blackAF     skip  
#ThatsHarassment          0     1           7    #ThatsHarassment     skip  
#MeToo, Now What?         0     1           7   #MeToo, Now What?     skip  

[5611 rows x 11 columns]

In [160]:
df.reset_index(inplace=True)

In [161]:
df.head()

Title  Year  Age  IMDb Rotten Tomatoes  Netflix  Hulu  \
0  頭文字D First Stage  1998   7+   8.4             NaN        0     1   
1               阳关道  2018  NaN   3.1             NaN        1     0   
2           那年花開月正圓  2017   7+   7.8             NaN        0     0   
3       恋と選挙とチョコレート  2012  18+   7.2             NaN        0     1   
4             Тобот  2010  NaN   6.2             NaN        1     0   

   Prime Video  Disney+  type  new_column duplicated_column todolist  
0            0        0     1           7  頭文字D First Stage     skip  
1            0        0     1           7               阳关道     skip  
2            1        0     1           7           那年花開月正圓     skip  
3            0        0     1           7       恋と選挙とチョコレート     skip  
4            1        0     1           7             Тобот     skip

## Get a sample

In [162]:
df2=df.sample(20)

In [163]:
df2

Title  Year  Age  IMDb  \
1092         The GiggleBellies - Monster Truck Learning  2017  NaN   NaN   
4990                                         Blown Away  2019   7+   6.8   
1084                                      The Gong Show  1976   7+   7.0   
2309                                         Our Planet  2019   7+   9.3   
2221                                       Pete the Cat  2017  all   7.5   
1922                                  Romanzo Criminale  2008  16+   8.6   
4431                             Descendants of the Sun  2016  16+   8.4   
2959                                         Lookalikes  2016  NaN   6.2   
2719                                      Metalocalypse  2006  18+   8.2   
3848                                   Girls und Panzer  2012  16+   7.5   
3000              Littlest Pet Shop: A World of Our Own  2018  all   6.4   
337                                               Valor  2017  16+   5.9   
3345  Is It Wrong to Try to Pick Up Girls in a Dungeon?  2015  16+   7.4   
5067                                    Big Easy Motors  2016   7+   6.2   
3651                                       Heartstrings  2011  16+   7.2   
1363                                Tell Me You Love Me  2007  18+   7.3   
4975                                           Board AF  2016  NaN   NaN   
2877                                    Lupin the Third  1971  16+   7.9   
2420                                        No Activity  2017  18+   6.7   
3906                               Gardens of the World  2007  NaN   NaN   

     Rotten Tomatoes  Netflix  Hulu  Prime Video  Disney+  type  new_column  \
1092             NaN        0     0            1        0     1           7   
4990             NaN        1     0            0        0     1           7   
1084             NaN        0     1            0        0     1           7   
2309             93%        1     0            0        0     1           7   
2221             NaN        0     0            1        0     1           7   
1922             NaN        0     0            1        0     1           7   
4431             NaN        1     1            1        0     1           7   
2959             NaN        1     0            0        0     1           7   
2719             NaN        0     1            0        0     1           7   
3848             NaN        1     1            0        0     1           7   
3000             NaN        1     0            0        0     1           7   
337              22%        1     0            0        0     1           7   
3345             NaN        1     1            0        0     1           7   
5067             NaN        0     0            1        0     1           7   
3651             NaN        0     1            0        0     1           7   
1363             62%        0     0            1        0     1           7   
4975             NaN        0     0            1        0     1           7   
2877             NaN        0     1            0        0     1           7   
2420             NaN        0     1            0        0     1           7   
3906             NaN        0     0            1        0     1           7   

                                      duplicated_column  todolist  
1092         The GiggleBellies - Monster Truck Learning      skip  
4990                                         Blown Away      skip  
1084                                      The Gong Show      skip  
2309                                         Our Planet  to watch  
2221                                       Pete the Cat      skip  
1922                                  Romanzo Criminale      skip  
4431                             Descendants of the Sun      skip  
2959                                         Lookalikes      skip  
2719                                      Metalocalypse      skip  
3848                                   Girls und Panzer      skip  
3000              Littlest Pet Shop: A World 

## Add rows

In [165]:
newdf=df.append(df2)
newdf

Title  Year  Age  IMDb Rotten Tomatoes  Netflix  Hulu  \
0         頭文字D First Stage  1998   7+   8.4             NaN        0     1   
1                      阳关道  2018  NaN   3.1             NaN        1     0   
2                  那年花開月正圓  2017   7+   7.8             NaN        0     0   
3              恋と選挙とチョコレート  2012  18+   7.2             NaN        0     1   
4                    Тобот  2010  NaN   6.2             NaN        1     0   
...                    ...   ...  ...   ...             ...      ...   ...   
1363   Tell Me You Love Me  2007  18+   7.3             62%        0     0   
4975              Board AF  2016  NaN   NaN             NaN        0     0   
2877       Lupin the Third  1971  16+   7.9             NaN        0     1   
2420           No Activity  2017  18+   6.7             NaN        0     1   
3906  Gardens of the World  2007  NaN   NaN             NaN        0     0   

      Prime Video  Disney+  type  new_column     duplicated_column todolist  
0               0        0     1           7      頭文字D First Stage     skip  
1               0        0     1           7                   阳关道     skip  
2               1        0     1           7               那年花開月正圓     skip  
3               0        0     1           7           恋と選挙とチョコレート     skip  
4               1        0     1           7                 Тобот     skip  
...           ...      ...   ...         ...                   ...      ...  
1363            1        0     1           7   Tell Me You Love Me     skip  
4975            1        0     1           7              Board AF     skip  
2877            0        0     1           7       Lupin the Third     skip  
2420            0        0     1           7           No Activity     skip  
3906            1        0     1           7  Gardens of the World     skip  

[5631 rows x 13 columns]

In [166]:
df.shape

(5611, 13)

In [167]:
newdf.shape

(5631, 13)

In [168]:
df.loc['crocodile',:]=['Random series', 2014, "all", 11,0,0,0,0,0,0,1,"","skip"]

In [169]:
df.tail()

Title    Year  Age  IMDb Rotten Tomatoes  Netflix  \
5607       (The Hook Up Plan)  2018.0  16+   7.4             NaN      1.0   
5608                 #blackAF  2020.0  18+   6.6             NaN      1.0   
5609         #ThatsHarassment  2018.0  NaN   NaN             NaN      0.0   
5610        #MeToo, Now What?  2018.0  NaN   6.6             NaN      0.0   
crocodile       Random series  2014.0  all  11.0               0      0.0   

           Hulu  Prime Video  Disney+  type  new_column   duplicated_column  \
5607        0.0          0.0      0.0   1.0         7.0  (The Hook Up Plan)   
5608        0.0          0.0      0.0   1.0         7.0            #blackAF   
5609        1.0          1.0      0.0   1.0         7.0    #ThatsHarassment   
5610        0.0          1.0      0.0   1.0         7.0   #MeToo, Now What?   
crocodile   0.0          0.0      0.0   0.0         1.0                       

          todolist  
5607          skip  
5608          skip  
5609          skip  
5610          skip  
crocodile     skip

# Drop rows

In [170]:
df.drop('crocodile').tail()

Title    Year  Age  IMDb Rotten Tomatoes  Netflix  Hulu  \
5606               .hack  2002.0   7+   6.9             18%      0.0   1.0   
5607  (The Hook Up Plan)  2018.0  16+   7.4             NaN      1.0   0.0   
5608            #blackAF  2020.0  18+   6.6             NaN      1.0   0.0   
5609    #ThatsHarassment  2018.0  NaN   NaN             NaN      0.0   1.0   
5610   #MeToo, Now What?  2018.0  NaN   6.6             NaN      0.0   0.0   

      Prime Video  Disney+  type  new_column   duplicated_column todolist  
5606          0.0      0.0   1.0         7.0               .hack     skip  
5607          0.0      0.0   1.0         7.0  (The Hook Up Plan)     skip  
5608          0.0      0.0   1.0         7.0            #blackAF     skip  
5609          1.0      0.0   1.0         7.0    #ThatsHarassment     skip  
5610          1.0      0.0   1.0         7.0   #MeToo, Now What?     skip

In [ ]:
df.loc["crocodile","IMDb"]="I'm an intruder"
df.tail()

# Duplicated rows

In [171]:
newdf.duplicated()

0       False
1       False
2       False
3       False
4       False
        ...  
1363     True
4975     True
2877     True
2420     True
3906     True
Length: 5631, dtype: bool

In [173]:
newdf.duplicated(subset=["Year"])

0       False
1       False
2       False
3       False
4       False
        ...  
1363     True
4975     True
2877     True
2420     True
3906     True
Length: 5631, dtype: bool

In [175]:
newdf.duplicated().sum()

20

In [176]:
newdf.drop_duplicates()

Title  Year  Age  IMDb Rotten Tomatoes  Netflix  Hulu  \
0       頭文字D First Stage  1998   7+   8.4             NaN        0     1   
1                    阳关道  2018  NaN   3.1             NaN        1     0   
2                那年花開月正圓  2017   7+   7.8             NaN        0     0   
3            恋と選挙とチョコレート  2012  18+   7.2             NaN        0     1   
4                  Тобот  2010  NaN   6.2             NaN        1     0   
...                  ...   ...  ...   ...             ...      ...   ...   
5606               .hack  2002   7+   6.9             18%        0     1   
5607  (The Hook Up Plan)  2018  16+   7.4             NaN        1     0   
5608            #blackAF  2020  18+   6.6             NaN        1     0   
5609    #ThatsHarassment  2018  NaN   NaN             NaN        0     1   
5610   #MeToo, Now What?  2018  NaN   6.6             NaN        0     0   

      Prime Video  Disney+  type  new_column   duplicated_column todolist  
0               0        0     1           7    頭文字D First Stage     skip  
1               0        0     1           7                 阳关道     skip  
2               1        0     1           7             那年花開月正圓     skip  
3               0        0     1           7         恋と選挙とチョコレート     skip  
4               1        0     1           7               Тобот     skip  
...           ...      ...   ...         ...                 ...      ...  
5606            0        0     1           7               .hack     skip  
5607            0        0     1           7  (The Hook Up Plan)     skip  
5608            0        0     1           7            #blackAF     skip  
5609            1        0     1           7    #ThatsHarassment     skip  
5610            1        0     1           7   #MeToo, Now What?     skip  

[5611 rows x 13 columns]

## Dropna

Droping every single row that have at least one single value in any column

In [178]:
df.dropna()

Title    Year  Age             IMDb  \
13                             mixed-ish  2019.0   7+              6.2   
14                               iZombie  2015.0  16+              7.9   
16                             black-ish  2014.0   7+              7.1   
23                                   Zoo  2015.0  16+              6.8   
27         Zoey's Extraordinary Playlist  2020.0  16+              8.1   
...                                  ...     ...  ...              ...   
5597                            11.22.63  2016.0  18+              8.2   
5602          10 Things I Hate About You  2009.0  16+              7.3   
5604                          09-01-2001  2018.0  16+              7.7   
5606                               .hack  2002.0   7+              6.9   
crocodile                  Random series  2014.0  all  I'm an intruder   

          Rotten Tomatoes  Netflix  Hulu  Prime Video  Disney+  type  \
13                    80%      0.0   1.0          0.0      0.0   1.0   
14                    91%      1.0   0.0          0.0      0.0   1.0   
16                    91%      0.0   1.0          0.0      0.0   1.0   
23                    82%      1.0   0.0          0.0      0.0   1.0   
27                    74%      0.0   1.0          0.0      0.0   1.0   
...                   ...      ...   ...          ...      ...   ...   
5597                  80%      0.0   1.0          0.0      0.0   1.0   
5602                  88%      0.0   1.0          0.0      0.0   1.0   
5604                  81%      0.0   1.0          0.0      0.0   1.0   
5606                  18%      0.0   1.0          0.0      0.0   1.0   
crocodile               0      0.0   0.0          0.0      0.0   0.0   

           new_column              duplicated_column todolist  
13                7.0                      mixed-ish     skip  
14                7.0                        iZombie     skip  
16                7.0                      black-ish     skip  
23                7.0                            Zoo     skip  
27                7.0  Zoey's Extraordinary Playlist     skip  
...               ...                            ...      ...  
5597              7.0                       11.22.63     skip  
5602              7.0     10 Things I Hate About You     skip  
5604              7.0                     09-01-2001     skip  
5606              7.0                          .hack     skip  
crocodile         1.0                                    skip  

[932 rows x 13 columns]

## ATTENTION

If I drop a column in a dataframe that has been assigned another dataframe, the column will be dropped in all the dataframes that are associated.

In [179]:
df1=df

In [180]:
df1.drop("todolist",axis=1, inplace=True)

In [181]:
df

Title    Year  Age             IMDb Rotten Tomatoes  \
0            頭文字D First Stage  1998.0   7+              8.4             NaN   
1                         阳关道  2018.0  NaN              3.1             NaN   
2                     那年花開月正圓  2017.0   7+              7.8             NaN   
3                 恋と選挙とチョコレート  2012.0  18+              7.2             NaN   
4                       Тобот  2010.0  NaN              6.2             NaN   
...                       ...     ...  ...              ...             ...   
5607       (The Hook Up Plan)  2018.0  16+              7.4             NaN   
5608                 #blackAF  2020.0  18+              6.6             NaN   
5609         #ThatsHarassment  2018.0  NaN              NaN             NaN   
5610        #MeToo, Now What?  2018.0  NaN              6.6             NaN   
crocodile       Random series  2014.0  all  I'm an intruder               0   

           Netflix  Hulu  Prime Video  Disney+  type  new_column  \
0              0.0   1.0          0.0      0.0   1.0         7.0   
1              1.0   0.0          0.0      0.0   1.0         7.0   
2              0.0   0.0          1.0      0.0   1.0         7.0   
3              0.0   1.0          0.0      0.0   1.0         7.0   
4              1.0   0.0          1.0      0.0   1.0         7.0   
...            ...   ...          ...      ...   ...         ...   
5607           1.0   0.0          0.0      0.0   1.0         7.0   
5608           1.0   0.0          0.0      0.0   1.0         7.0   
5609           0.0   1.0          1.0      0.0   1.0         7.0   
5610           0.0   0.0          1.0      0.0   1.0         7.0   
crocodile      0.0   0.0          0.0      0.0   0.0         1.0   

            duplicated_column  
0            頭文字D First Stage  
1                         阳关道  
2                     那年花開月正圓  
3                 恋と選挙とチョコレート  
4                       Тобот  
...                       ...  
5607       (The Hook Up Plan)  
5608                 #blackAF  
5609         #ThatsHarassment  
5610        #MeToo, Now What?  
crocodile                      

[5612 rows x 12 columns]

## Copy

In [191]:
df2=df.copy()

In [193]:
df2.drop('Title',axis=1)
df2

Title    Year  Age             IMDb Rotten Tomatoes  \
0            頭文字D First Stage  1998.0   7+              8.4             NaN   
1                         阳关道  2018.0  NaN              3.1             NaN   
2                     那年花開月正圓  2017.0   7+              7.8             NaN   
3                 恋と選挙とチョコレート  2012.0  18+              7.2             NaN   
4                       Тобот  2010.0  NaN              6.2             NaN   
...                       ...     ...  ...              ...             ...   
5607       (The Hook Up Plan)  2018.0  16+              7.4             NaN   
5608                 #blackAF  2020.0  18+              6.6             NaN   
5609         #ThatsHarassment  2018.0  NaN              NaN             NaN   
5610        #MeToo, Now What?  2018.0  NaN              6.6             NaN   
crocodile       Random series  2014.0  all  I'm an intruder               0   

           Netflix  Hulu  Prime Video  Disney+  type  new_column  \
0              0.0   1.0          0.0      0.0   1.0         7.0   
1              1.0   0.0          0.0      0.0   1.0         7.0   
2              0.0   0.0          1.0      0.0   1.0         7.0   
3              0.0   1.0          0.0      0.0   1.0         7.0   
4              1.0   0.0          1.0      0.0   1.0         7.0   
...            ...   ...          ...      ...   ...         ...   
5607           1.0   0.0          0.0      0.0   1.0         7.0   
5608           1.0   0.0          0.0      0.0   1.0         7.0   
5609           0.0   1.0          1.0      0.0   1.0         7.0   
5610           0.0   0.0          1.0      0.0   1.0         7.0   
crocodile      0.0   0.0          0.0      0.0   0.0         1.0   

            duplicated_column  
0            頭文字D First Stage  
1                         阳关道  
2                     那年花開月正圓  
3                 恋と選挙とチョコレート  
4                       Тобот  
...                       ...  
5607       (The Hook Up Plan)  
5608                 #blackAF  
5609         #ThatsHarassment  
5610        #MeToo, Now What?  
crocodile                      

[5612 rows x 12 columns]

## Fillna

In [196]:
df.IMDb.fillna(0)

0                        8.4
1                        3.1
2                        7.8
3                        7.2
4                        6.2
                  ...       
5607                     7.4
5608                     6.6
5609                       0
5610                     6.6
crocodile    I'm an intruder
Name: IMDb, Length: 5612, dtype: object

In [198]:
df.IMDb.fillna("unknown")

0                        8.4
1                        3.1
2                        7.8
3                        7.2
4                        6.2
                  ...       
5607                     7.4
5608                     6.6
5609                 unknown
5610                     6.6
crocodile    I'm an intruder
Name: IMDb, Length: 5612, dtype: object

In [200]:
df.Age.fillna(df.Age.mode()[0])

0             7+
1            16+
2             7+
3            18+
4            16+
            ... 
5607         16+
5608         18+
5609         16+
5610         16+
crocodile    all
Name: Age, Length: 5612, dtype: object

## Complex filter

In [201]:
df[df.Year==2020]

Title    Year  Age IMDb  \
27              Zoey's Extraordinary Playlist  2020.0  16+  8.1   
31                               ZeroZeroZero  2020.0  18+  8.2   
144                Win the Wilderness: Alaska  2020.0  NaN  7.1   
184             Who Wants to Be a Millionaire  2020.0  16+  6.1   
193                               White Lines  2020.0  18+  6.2   
...                                       ...     ...  ...  ...   
5493                         AJ and the Queen  2020.0  NaN  7.6   
5502               A3! Season Spring & Summer  2020.0  16+  5.6   
5546                         9-1-1: Lone Star  2020.0  16+  6.4   
5600  100 Humans. Life's Questions. Answered.  2020.0  NaN  5.6   
5608                                 #blackAF  2020.0  18+  6.6   

     Rotten Tomatoes  Netflix  Hulu  Prime Video  Disney+  type  new_column  \
27               74%      0.0   1.0          0.0      0.0   1.0         7.0   
31               80%      0.0   0.0          1.0      0.0   1.0         7.0   
144              NaN      1.0   0.0          0.0      0.0   1.0         7.0   
184              NaN      0.0   1.0          0.0      0.0   1.0         7.0   
193              58%      1.0   0.0          0.0      0.0   1.0         7.0   
...              ...      ...   ...          ...      ...   ...         ...   
5493             50%      1.0   0.0          0.0      0.0   1.0         7.0   
5502             NaN      0.0   1.0          0.0      0.0   1.0         7.0   
5546             NaN      0.0   1.0          0.0      0.0   1.0         7.0   
5600             NaN      1.0   0.0          0.0      0.0   1.0         7.0   
5608             NaN      1.0   0.0          0.0      0.0   1.0         7.0   

                            duplicated_column  
27              Zoey's Extraordinary Playlist  
31                               ZeroZeroZero  
144                Win the Wilderness: Alaska  
184             Who Wants to Be a Millionaire  
193                               White Lines  
...                                       ...  
5493                         AJ and the Queen  
5502               A3! Season Spring & Summer  
5546                         9-1-1: Lone Star  
5600  100 Humans. Life's Questions. Answered.  
5608                                 #blackAF  

[177 rows x 12 columns]

In [203]:
df.drop('crocodile',inplace=True)

In [205]:
df[df.IMDb>0]

Title    Year  Age IMDb Rotten Tomatoes  Netflix  Hulu  \
0       頭文字D First Stage  1998.0   7+  8.4             NaN      0.0   1.0   
1                    阳关道  2018.0  NaN  3.1             NaN      1.0   0.0   
2                那年花開月正圓  2017.0   7+  7.8             NaN      0.0   0.0   
3            恋と選挙とチョコレート  2012.0  18+  7.2             NaN      0.0   1.0   
4                  Тобот  2010.0  NaN  6.2             NaN      1.0   0.0   
...                  ...     ...  ...  ...             ...      ...   ...   
5604          09-01-2001  2018.0  16+  7.7             81%      0.0   1.0   
5606               .hack  2002.0   7+  6.9             18%      0.0   1.0   
5607  (The Hook Up Plan)  2018.0  16+  7.4             NaN      1.0   0.0   
5608            #blackAF  2020.0  18+  6.6             NaN      1.0   0.0   
5610   #MeToo, Now What?  2018.0  NaN  6.6             NaN      0.0   0.0   

      Prime Video  Disney+  type  new_column   duplicated_column  
0             0.0      0.0   1.0         7.0    頭文字D First Stage  
1             0.0      0.0   1.0         7.0                 阳关道  
2             1.0      0.0   1.0         7.0             那年花開月正圓  
3             0.0      0.0   1.0         7.0         恋と選挙とチョコレート  
4             1.0      0.0   1.0         7.0               Тобот  
...           ...      ...   ...         ...                 ...  
5604          0.0      0.0   1.0         7.0          09-01-2001  
5606          0.0      0.0   1.0         7.0               .hack  
5607          0.0      0.0   1.0         7.0  (The Hook Up Plan)  
5608          0.0      0.0   1.0         7.0            #blackAF  
5610          1.0      0.0   1.0         7.0   #MeToo, Now What?  

[4450 rows x 12 columns]

In [206]:
(df.Year==2020)&(df.IMDb>0)

0       False
1       False
2       False
3       False
4       False
        ...  
5606    False
5607    False
5608     True
5609    False
5610    False
Length: 5611, dtype: bool

In [207]:
df[(df.Year==2020)&(df.IMDb>0)]

Title    Year  Age IMDb  \
27              Zoey's Extraordinary Playlist  2020.0  16+  8.1   
31                               ZeroZeroZero  2020.0  18+  8.2   
144                Win the Wilderness: Alaska  2020.0  NaN  7.1   
184             Who Wants to Be a Millionaire  2020.0  16+  6.1   
193                               White Lines  2020.0  18+  6.2   
...                                       ...     ...  ...  ...   
5493                         AJ and the Queen  2020.0  NaN  7.6   
5502               A3! Season Spring & Summer  2020.0  16+  5.6   
5546                         9-1-1: Lone Star  2020.0  16+  6.4   
5600  100 Humans. Life's Questions. Answered.  2020.0  NaN  5.6   
5608                                 #blackAF  2020.0  18+  6.6   

     Rotten Tomatoes  Netflix  Hulu  Prime Video  Disney+  type  new_column  \
27               74%      0.0   1.0          0.0      0.0   1.0         7.0   
31               80%      0.0   0.0          1.0      0.0   1.0         7.0   
144              NaN      1.0   0.0          0.0      0.0   1.0         7.0   
184              NaN      0.0   1.0          0.0      0.0   1.0         7.0   
193              58%      1.0   0.0          0.0      0.0   1.0         7.0   
...              ...      ...   ...          ...      ...   ...         ...   
5493             50%      1.0   0.0          0.0      0.0   1.0         7.0   
5502             NaN      0.0   1.0          0.0      0.0   1.0         7.0   
5546             NaN      0.0   1.0          0.0      0.0   1.0         7.0   
5600             NaN      1.0   0.0          0.0      0.0   1.0         7.0   
5608             NaN      1.0   0.0          0.0      0.0   1.0         7.0   

                            duplicated_column  
27              Zoey's Extraordinary Playlist  
31                               ZeroZeroZero  
144                Win the Wilderness: Alaska  
184             Who Wants to Be a Millionaire  
193                               White Lines  
...                                       ...  
5493                         AJ and the Queen  
5502               A3! Season Spring & Summer  
5546                         9-1-1: Lone Star  
5600  100 Humans. Life's Questions. Answered.  
5608                                 #blackAF  

[171 rows x 12 columns]

## isin

In [208]:
df.Age.value_counts()

16+    1018
7+      848
18+     750
all     545
13+       4
Name: Age, dtype: int64

How many films my little sister of 12 yo can watch ?
- all: 545
- 7+ 848

So my little sister can watch 1393 films.

How to display them ?

In [211]:
df[(df.Age=="7+")|(df.Age=="all")]

Title    Year Age IMDb Rotten Tomatoes  Netflix  \
0             頭文字D First Stage  1998.0  7+  8.4             NaN      0.0   
2                      那年花開月正圓  2017.0  7+  7.8             NaN      0.0   
13                   mixed-ish  2019.0  7+  6.2             80%      0.0   
16                   black-ish  2014.0  7+  7.1             91%      0.0   
18       Zumbo's Just Desserts  2016.0  7+  6.9             NaN      1.0   
...                        ...     ...  ..  ...             ...      ...   
5569  3Below: Tales of Arcadia  2018.0  7+  7.8            100%      1.0   
5572                 30 for 30  2009.0  7+  8.8             NaN      1.0   
5574                        3%  2016.0  7+  7.4             33%      1.0   
5583                  205 Live  2016.0  7+  7.1             NaN      0.0   
5606                     .hack  2002.0  7+  6.9             18%      0.0   

      Hulu  Prime Video  Disney+  type  new_column         duplicated_column  
0      1.0          0.0      0.0   1.0         7.0          頭文字D First Stage  
2      0.0          1.0      0.0   1.0         7.0                   那年花開月正圓  
13     1.0          0.0      0.0   1.0         7.0                 mixed-ish  
16     1.0          0.0      0.0   1.0         7.0                 black-ish  
18     0.0          0.0      0.0   1.0         7.0     Zumbo's Just Desserts  
...    ...          ...      ...   ...         ...                       ...  
5569   0.0          0.0      0.0   1.0         7.0  3Below: Tales of Arcadia  
5572   0.0          0.0      0.0   1.0         7.0                 30 for 30  
5574   0.0          0.0      0.0   1.0         7.0                        3%  
5583   1.0          0.0      0.0   1.0         7.0                  205 Live  
5606   1.0          0.0      0.0   1.0         7.0                     .hack  

[1393 rows x 12 columns]

In [212]:
df[df.Age.isin(["7+","all"])]

Title    Year Age IMDb Rotten Tomatoes  Netflix  \
0             頭文字D First Stage  1998.0  7+  8.4             NaN      0.0   
2                      那年花開月正圓  2017.0  7+  7.8             NaN      0.0   
13                   mixed-ish  2019.0  7+  6.2             80%      0.0   
16                   black-ish  2014.0  7+  7.1             91%      0.0   
18       Zumbo's Just Desserts  2016.0  7+  6.9             NaN      1.0   
...                        ...     ...  ..  ...             ...      ...   
5569  3Below: Tales of Arcadia  2018.0  7+  7.8            100%      1.0   
5572                 30 for 30  2009.0  7+  8.8             NaN      1.0   
5574                        3%  2016.0  7+  7.4             33%      1.0   
5583                  205 Live  2016.0  7+  7.1             NaN      0.0   
5606                     .hack  2002.0  7+  6.9             18%      0.0   

      Hulu  Prime Video  Disney+  type  new_column         duplicated_column  
0      1.0          0.0      0.0   1.0         7.0          頭文字D First Stage  
2      0.0          1.0      0.0   1.0         7.0                   那年花開月正圓  
13     1.0          0.0      0.0   1.0         7.0                 mixed-ish  
16     1.0          0.0      0.0   1.0         7.0                 black-ish  
18     0.0          0.0      0.0   1.0         7.0     Zumbo's Just Desserts  
...    ...          ...      ...   ...         ...                       ...  
5569   0.0          0.0      0.0   1.0         7.0  3Below: Tales of Arcadia  
5572   0.0          0.0      0.0   1.0         7.0                 30 for 30  
5574   0.0          0.0      0.0   1.0         7.0                        3%  
5583   1.0          0.0      0.0   1.0         7.0                  205 Live  
5606   1.0          0.0      0.0   1.0         7.0                     .hack  

[1393 rows x 12 columns]